### PR TITLE
chore: upgrade deps

### PR DESCRIPTION
**Pull request by jasoniangreen**
See inline comments for an explanation of each config change. In general these choices were made to give the most protection and ensuring best practices while avoiding large changes in the AJV codebase.

You can see the result of these rules in the [AJV PR](https://github.com/ajv-validator/ajv/pull/2396/files).

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -11,15 +11,20 @@ module.exports = {
     "eslint:recommended",
     "plugin:@typescript-eslint/recommended",
     "plugin:@typescript-eslint/recommended-requiring-type-checking",
-    "prettier/@typescript-eslint",
+    "prettier",
   ],
   rules: {
     ...jsConfig.rules,
+    "@typescript-eslint/no-unsafe-argument": "off", // 123 instances
```
Comment (jasoniangreen) [inline .eslintrc.js]:
```suggestion
    "@typescript-eslint/no-unsafe-argument": "off",
```

File: .eslintrc.js [hunk 2]
```diff
@@ -11,15 +11,20 @@ module.exports = {
     "eslint:recommended",
     "plugin:@typescript-eslint/recommended",
     "plugin:@typescript-eslint/recommended-requiring-type-checking",
-    "prettier/@typescript-eslint",
+    "prettier",
   ],
   rules: {
     ...jsConfig.rules,
+    "@typescript-eslint/no-unsafe-argument": "off",
```
Comment (jasoniangreen) [inline .eslintrc.js:18]: This appears to be a new rule that didn't exist before. We had 123 violations which is too much to fix or disable by line.

File: .eslintrc.js [hunk 3]
```diff
@@ -11,15 +11,20 @@ module.exports = {
...
     "@typescript-eslint/array-type": "error",
     "@typescript-eslint/consistent-type-assertions": "error",
     "@typescript-eslint/consistent-type-definitions": "error",
+    "@typescript-eslint/no-redundant-type-constituents": "error",
```
Comment (jasoniangreen) [inline .eslintrc.js:22]: New rule that is on by default it seems, so I added it to be explicit as it seems useful.

File: .eslintrc.js [hunk 4]
```diff
@@ -11,15 +11,20 @@ module.exports = {
...
+    "@typescript-eslint/no-shadow": "error",
```
Comment (jasoniangreen) [inline .eslintrc.js:23]: Ported from the js ruleset

File: .eslintrc.js [hunk 5]
```diff
@@ -11,15 +11,20 @@ module.exports = {
...
+    "@typescript-eslint/no-redeclare": "error",
```
Comment (jasoniangreen) [inline .eslintrc.js:24]: Ported from the js ruleset

File: .eslintrc.js [hunk 6]
```diff
@@ -11,15 +11,20 @@ module.exports = {
...
+    "@typescript-eslint/no-base-to-string": ["error", { ignoredTypeNames: ["Error", "RegExp", "RegExpLike", "URL", "URLSearchParams"] }], // 1 instance
```
Comment (jasoniangreen) [inline .eslintrc.js:25]: A new rule added by default. I have added some standard exceptions though the one that actually triggered in ajv was `RegExpLike`.

File: .eslintrc.js [hunk 7]
```diff
@@ -11,15 +11,20 @@ module.exports = {
...
+    "@typescript-eslint/no-use-before-define": ["error", "nofunc"],
```
Comment (jasoniangreen) [inline .eslintrc.js:26]: Ported from the js ruleset

File: .eslintrc.js [hunk 8]
```diff
@@ -29,7 +34,7 @@ module.exports = {
     "no-invalid-this": "off",
     "@typescript-eslint/no-invalid-this": "error",
     "@typescript-eslint/no-misused-new": "error",
-    "@typescript-eslint/no-parameter-properties": "error",
+    "@typescript-eslint/parameter-properties": ["error", {prefer: "parameter-property"}],
```
Comment (jasoniangreen) [inline .eslintrc.js:37]: This way this rule works has been updated.

File: .eslintrc_js.js [hunk 9]
```diff
@@ -25,14 +25,12 @@ module.exports = {
...
     "no-path-concat": "error",
-    "no-redeclare": "error",
     "no-return-assign": "error",
     "no-sequences": "error",
-    "no-shadow": "warn",
     "no-template-curly-in-string": "error",
     "no-trailing-spaces": "error",
     "no-undef-init": "error",
-    "no-use-before-define": ["error", "nofunc"],
+    "no-use-before-define": "off",
```
Comment (jasoniangreen) [inline .eslintrc_js.js:33]: Must be off to enable the TS version.

File: package.json [hunk 10]
```diff
@@ -21,14 +21,14 @@
...
-    "prettier": "^2.0.5",
-    "typescript": "^3.9.7"
+    "@typescript-eslint/eslint-plugin": "^7.3.1",
+    "@typescript-eslint/parser": "^7.3.1",
+    "eslint": "^8.57.0",
+    "eslint-config-prettier": "^9.1.0",
+    "husky": "^9.0.11",
+    "lint-staged": "^15.2.2",
+    "prettier": "3.0.3",
+    "typescript": "^4.9.5"
```
Comment (jasoniangreen) [inline package.json:31]: Have updated TS to the same version of the main repo.

**Review by epoberezkin — APPROVED** [review]

**Pull request merged**